src/main/scala/beam/agentsim/agents/choice/logit/MultinomialLogit.scala

```diff
@@ -1,146 +1,170 @@
   */
-case class MultinomialLogit(alternativeParams: Map[String, AlternativeParams]) extends LazyLogging {
-
+class MultinomialLogit[A, T](
+  val utilityFunctions: Map[A, Map[T, UtilityFunctionOperation]],
+  common: Map[T, UtilityFunctionOperation]
+) extends LazyLogging {
+
+  /**
+    * Sample over a set of types A by calculating the probabilities of each alternative
+    * and then draw one randomly.
+    *
+    * For details see page 103, formula 5.8 in
+    * Ben-Akiva, M., & Lerman, S. R. (1994). Discrete choice analysis : theory and application to travel demand. 6th print. Cambridge (Mass.): MIT press.
+    *
+    * @param alternatives the alternatives that should be sampled
+    * @param random a random we can sample on
+    * @return
+    */
   def sampleAlternative(
-    alternatives: Vector[AlternativeAttributes],
+    alternatives: Map[A, Map[T, Double]],
     random: Random
-  ): Option[String] = {
-    val expV = alternatives.map(alt => Math.exp(getUtilityOfAlternative(alt)))
-    // If any is +Inf then choose that as the certain alternative
-    val indsOfPosInf = for (theExpV <- expV.zipWithIndex if theExpV._1 == Double.PositiveInfinity)
-      yield theExpV._2
-    if (indsOfPosInf.nonEmpty) {
-      // Take the first
-      Some(alternatives(indsOfPosInf.head).alternativeName)
-    } else {
-      val sumExpV = expV.sum
-      val cumulProbs = expV.map(_ / sumExpV).scanLeft(0.0)(_ + _).zipWithIndex
-      val randDraw = random.nextDouble()
-      val idxAboveDraw = for (prob <- cumulProbs if prob._1 > randDraw) yield prob._2
-      if (idxAboveDraw.isEmpty) {
-        None
-      } else {
-        val chosenIdx = idxAboveDraw.head - 1
-        Some(alternatives(chosenIdx).alternativeName)
-      }
-    }
-  }
-
-  def getExpectedMaximumUtility(alternatives: Vector[AlternativeAttributes]): Double = {
-    Math.log(alternatives.map(alt => Math.exp(getUtilityOfAlternative(alt))).sum)
-  }
-
-  def getUtilityOfAlternative(alternative: AlternativeAttributes): Double = {
-    if (!alternativeParams.contains(alternative.alternativeName)) {
-      -1E100
-    } else {
-      (alternativeParams.getOrElse("COMMON", AlternativeParams.empty).params ++ alternativeParams(
-        alternative.alternativeName
-      ).params)
-        .map { theParam =>
-          if (alternative.attributes.contains(theParam._1)) {
-            theParam._2.paramType match {
-              case Multiplier =>
-                (theParam._2.paramValue * alternative.attributes(theParam._1)).toDouble
-              case Intercept =>
-                theParam._2.paramValue.toDouble
+  ): Option[MultinomialLogit.MNLSample[A]] = {
+    if (alternatives.isEmpty) None
+    else {
+
+      // one-pass evaluation and sorting in descending order by alternative utilities
+      val altsWithUtilitySortedDesc: Iterable[(A, Double)] =
+        alternatives.foldLeft(SortedSet.empty[(A, Double)](Ordering.by { -_._2 })) {
```

Comment:
We MAY want to be more explicit in our ordering as we've had some performance issues on the way generic ordering works. Thoughts @REASY?